pkg/util/tracing.go

```diff
@@ -0,0 +1,96 @@
+package util
+
+import (
+	"sync"
+	"time"
+
+	"go.opencensus.io/trace"
+)
+
+// ConstantRateTraceSampler uses a token bucket algorithm to sample traces. This is useful as an alternative
+// to probability-based sampling when the service has a low number of requests per second.
+//
+// Inspired in part by https://github.com/jaegertracing/jaeger-client-go/blob/master/utils/rate_limiter.go
+
+type ConstantRateTraceSampler struct {
+	lock sync.Mutex
+
+	// The number of available tokens in the bucket.
+	AvailableTokens int64
```

Comment:
I think that all of these fields may remain private.

I would also suggest to group these fields by locking requirement. The general style in the Buildbarn codebase is to put constant fields at the very top of the struct, followed by placing all fields that require locking right underneath the lock declaration. That way it becomes self-documenting what the locking pattern is.

Reply:
> I think that all of these fields may remain private.

These fields were private in an earlier iteration of this change (in a version before the one seen in this PR). I switched the fields to public as, because the test is in the `util_test` package and this code is in the `util` package, the test originally constructed the `constantRateTraceSampler` struct directly in order to control the clock mock. It had to do this because the code originally returned the `sample` method (and not the struct).

I then switched to returning the struct and not the `Sample` method from the construction method so that the test could avoid needing to construct the struct directly, but that in turn still required the field to be public for the test to be able to control the clock mock.

So the weird structure of this code is due to mocking the clock for the test. Your comment to use `NewMockClock` seems like it could potentially remove the need for these "visibility gymnastics."

Reply:
> I would also suggest to group these fields by locking requirement. The general style in the Buildbarn codebase is to put constant fields at the very top of the struct, followed by placing all fields that require locking right underneath the lock declaration. That way it becomes self-documenting what the locking pattern is.

Will do.